File: Java-Core/src/main/java/prompto/type/RangeType.java
package prompto.type;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;

import prompto.intrinsic.PromptoRange;
import prompto.runtime.Context;

public class RangeType extends ContainerType {
	
	public RangeType(IType itemType) {
		super(Family.RANGE, itemType, itemType.getTypeName()+"[..]");
	}
	
	@Override
	public boolean isAssignableTo(Context context, IType other) {
		return this.equals(other);
	}
	
	static Map<IType,Class<?>> rangeClassMap = createRangeClassMap();
	
	private static Map<IType, Class<?>> createRangeClassMap() {
		Map<IType,Class<?>> map = new HashMap<>();
		map.put(CharacterType.instance(), PromptoRange.Character.class);
		map.put(IntegerType.instance(), PromptoRange.Long.class);
		map.put(DateType.instance(), PromptoRange.Date.class);
		map.put(TimeType.instance(), PromptoRange.Time.class);
		return map;
	}
	
	@Override
	public Type getJavaType() {
		return rangeClassMap.get(itemType); 
	}
	

	@Override
	public boolean equals(Object obj) {
		if(obj==this)
			return true; 
		if(obj==null)
			return false;
		if(!(obj instanceof RangeType))
			return false;
		RangeType other = (RangeType)obj;
		return this.getItemType().equals(other.getItemType());
	}
	
	@Override
	public IType checkItem(Context context, IType other) {
		if(other==IntegerType.instance())
			return itemType;
		else
			return super.checkItem(context,other);
	}
	
	@Override
	public IType checkSlice(Context context) {
		return this;
	}
	
	@Override
	public IType checkIterator(Context context) {
		return itemType;
	}

	
}

File: Java-CodeServer/src/main/resources/js/prompto/type/RangeType.js
var CollectionType = require("./CollectionType").CollectionType;
var IntegerType = null;

exports.resolve = function() {
	IntegerType = require("./IntegerType").IntegerType;
};

function RangeType(itemType) {
    CollectionType.call(this, itemType.name+"[..]",itemType);
	return this;
}

RangeType.prototype = Object.create(CollectionType.prototype);
RangeType.prototype.constructor = RangeType;

/*
@Override
public boolean isAssignableTo(Context context, IType other) {
	return this.equals(other);
}

@Override
public Class<?> toJavaClass() {
	return null; // no equivalent
}


@Override
public boolean equals(Object obj) {
	if(obj==this)
		return true;
	if(obj==null)
		return false;
	if(!(obj instanceof RangeType))
		return false;
	RangeType other = (RangeType)obj;
	return this.getItemType().equals(other.getItemType());
}

*/

RangeType.prototype.checkItem = function(context, other) {
	if (other == IntegerType.instance) {
		return this.itemType;
	} else {
		return CollectionType.prototype.checkItem.call(this, context, other);
	}
};

RangeType.prototype.checkSlice = function(context) {
	return this;
};

RangeType.prototype.checkIterator = function(context) {
	return this.itemType;
};

exports.RangeType = RangeType;